### PR TITLE
Implement OASIS file writer

### PR DESCRIPTION
**Pull request by c1ydehhx**
In this PR, I added OASIS device. This device support writer only currently. This writer implement manhattan delta format by OASIS spec.

For example, if we expected +340 for N/S direction, it will be encoded with 0xD0 0x0A0xA8 0x05, and -340 will be 0xD2 0x0A0xA9 0x05.

I added some test to make sure it work properly, and I add some edge case to make sure it will not buggy when met edge case.

---

Updated (2026/02/19):

After the discussion with @yungyuc, I added 3 classes `modmesh::OasisWriter`, `modmesh::OasisRecordRect`, `modmesh::OasisRecordPoly` to output OASIS-format bytes.

The two record classes convert geometry information into OASIS-format bytes:
 - `OasisRecordPoly` class accepts coordinate pairs (e.g., `[[410, 720], [410, 920], [70, 920], ...]`) and converts them into OASIS-format bytes.
 - `OasisRecordRect` class accepts a lower-left coordinate pair (e.g., `[70, 800]`), width, height and converts them into OASIS-format bytes.

With these record classes, `OasisWriter` accepts `OasisRecordPoly` and `OasisRecordRect` objects and integrates these records with the minimal required headers (e.g., magic-bytes, START, END, CELL, and CELLNAME records) and outputs OASIS-format bytes. The output bytes should be loadable by third-party software.

These classes above are wrapped as Python classes via pybind11. In this PR, I added a few unit tests to verify the classes work properly:

 - `OasisRecordPoly` and `OasisRecordRect`: verify that the produced OASIS-format bytes are correct.
 - `OasisWriter`: verify the output bytes contain the required headers with integrated polygon and rectangle records.
 
In addition, I tested the output bytes manually with [KLayout software](https://www.klayout.de/). The output bytes from `OasisWriter` is accepted by KLayout, as shown below.

image

### CODE REVIEW

File: modmesh/plot/plane_layer.py [hunk 1]
```diff
@@ -25,6 +25,35 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 
+class ManhattanDeltaCodec:
```
Comment (yungyuc) [inline modmesh/plot/plane_layer.py]: Write I/O code in C++.  Python is too slow to be taken seriously for I/O.

Name it `modmesh::OasisDevice` for now.

File: tests/test_plane_layer.py [hunk 2]
```diff
@@ -29,6 +29,49 @@
 from modmesh.plot import plane_layer
 
 
+class ManhattanDeltaCodecTC(unittest.TestCase):
```
Comment (yungyuc) [inline tests/test_plane_layer.py]: Unit tests for the OASIS I/O should be in Python.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Points to address:

- [x] Write OASIS I/O code in C++.  Name it `modmesh::OasisDevice`.  Wrap it to Python for testing.
- [x] Manually verify the written file by loading it using a third-party viewer and paste the screenshot.

**Comment by c1ydehhx** [comment]
Hi @yungyuc,
So far away, I already migrate Manhatten 1-delta writer (Named it OasisWriter) from Python to C++, and use pybind11 to wrap it as a module.

>  Write OASIS I/O code in C++. Name it modmesh::OasisDevice. Wrap it to Python for testing.

Done.

> Manually verify the written file by loading it using a third-party viewer and paste the screenshot.

I'm concern that it will have too much diff in this PR if we want to output OASIS file. Althought we have geometry record now, we don't have some necessary OASIS header like magic bytes, START and END.

I would like to confirm that this PR should support OASIS file output or we can just test Manhatten 1-delta work properly with unit tests?

**Comment by yungyuc** [comment]
Linter did not pass.  Please fix code format and then request for review again.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/base.hpp>
+#include <pybind11/stl.h>
+
+#include <cstdint>
+#include <vector>
+
+using bytes = std::vector<uint8_t>;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Do not make an intrusive alias like this.  It is horrible to maintain.

There is nothing wrong with using `std::vector<uint8)_t>`?  Just spell out the type.

It is horrible to add an alias in the global name space.  Unacceptable almost always.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+namespace oasis
+{
+
+class OasisDevice
+{
+private:
+    static void append_1_delta(bytes & segment, int value);
+
+public:
+    OasisDevice() = default;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Rule of 5.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    static bytes writer(std::vector<int32_t> & values);
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Name it `write`, not `writer`, which is not a verb.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Some very basic points to address:

- [x] Do not make an intrusive alias in global namespace and for STL things like `std::vector<uint8)_t>`.
- [x] Rule of 5 for `OasisDevice`.
- [x] Name as `OasisDevice::write`, not `writer`, which is not a verb.

**Comment by yungyuc** [comment]
@tigercosmos @KHLee529 please help review

File: cpp/modmesh/oasis/pymod/wrap_oasis.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+#include <modmesh/oasis/oasis_device.hpp>
+
+namespace modmesh
+{
+
+namespace python
+{
+
+using namespace modmesh::oasis; // NOLINT(google-build-using-namespace)
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/pymod/wrap_oasis.cpp]: I recommend remove this.  Namespace is important for reading the code.

File: cpp/modmesh/oasis/oasis_device.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace modmesh
+{
+
+namespace oasis
+{
+
+void OasisDevice::append_1_delta(std::vector<uint8_t> & segment, int value)
+{
+    int dir_bit = value < 0 ? 1 : 0;
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/oasis_device.cpp]: Please add `const` for all the places if applicable.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+#include <cstdint>
+#include <vector>
+
+namespace modmesh
+{
+
+namespace oasis
+{
+
+class OasisDevice
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/oasis_device.hpp:53]: Please add document comments for classes and functions.

File: cpp/modmesh/oasis/oasis_device.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+    while (payload >= 1)
+    {
+        int first_bit = payload >= 128 ? 1 : 0;
+        segment.push_back((first_bit << 7) | (payload % 128));
+        payload /= 128;
+    }
+}
+
+std::vector<uint8_t> OasisDevice::write(std::vector<int32_t> & values)
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/oasis_device.cpp]: `const std::vector<int32_t> & values`

File: cpp/modmesh/oasis/oasis_device.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+#include <modmesh/oasis/oasis_device.hpp>
+#include <cstdint>
+
+namespace modmesh
+{
+
+namespace oasis
+{
+
+void OasisDevice::append_1_delta(std::vector<uint8_t> & segment, int value)
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/oasis_device.cpp]: This can be a local function in cpp only.

File: cpp/modmesh/oasis/oasis_device.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    if (payload == 0)
+    {
+        segment.push_back(0x02);
+    }
+
+    while (payload >= 1)
+    {
+        int first_bit = payload >= 128 ? 1 : 0;
+        segment.push_back((first_bit << 7) | (payload % 128));
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/oasis_device.cpp:83]: Please add comments to explain the algorithm. Including source doucment is even better.

File: modmesh/plot/plane_layer.py (hunk 1 above)
Comment (tigercosmos) [inline modmesh/plot/plane_layer.py]: Please remove it.

File: tests/test_oasis_device.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        device = modmesh.OasisDevice()
+        seg = device.writer([-340])
+
+        self.assertEqual(seg, [0xD2, 0x0A])
+
+    def test_with_integers(self):
+        device = modmesh.OasisDevice()
+        seg = device.writer([+340, +200, -340, -40, +300, -120, -300])
+
+        self.assertEqual(seg, [0xD0, 0x0A, 0xA0, 0x06, 0xD2,
```
Comment (tigercosmos) [inline tests/test_oasis_device.py]: It will be nice if you can explain what you try to test in the comments.

File: cpp/modmesh/oasis/pymod/wrap_oasis.cpp (hunk 6 above)
Comment (yungyuc) [reply cpp/modmesh/oasis/pymod/wrap_oasis.cpp]: Good catch.  `modmesh::oasis` namespace is not necessary and should be removed.

And, never using namespace unless it is in a very limited scope and can resolve extreme pain.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+#include <modmesh/base.hpp>
+#include <pybind11/stl.h>
+
+#include <cstdint>
+#include <vector>
+
+namespace modmesh
+{
+
+namespace oasis
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Do not use namespace `modmesh::oasis`.

File: tests/test_oasis_device.py (hunk 12 above)
Comment (yungyuc) [reply tests/test_oasis_device.py]: Use code comments to explain.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Points to address:

- [x] Do not use namespace `modmesh::oasis`.
- [x] Add very concise code comments for the key points in unit tests.

**Comment by c1ydehhx** [comment]
Hi @yungyuc and @tigercosmos.

Thanks for the review. 

I complete OASIS device to store polygon and rectangle and write as OASIS format. The OASIS format will need magic-byte, START record, END record, CELLNAME record and CELL record to work. For each record, I add [comment to describe its format and how to find the record spec in OASIS spec draft](https://github.com/solvcon/modmesh/pull/674/changes#diff-80d53e0d8adc09a351b6d83e630454c2d6414fe7a17a38bfed1a150ff10f95d8). I also add some comment to describe how unsigned and signed value work in OASIS. Please refer the comment in `oasis_device.cpp`. For the comment of tests. I leave a simple commant to let viewer refer the comment in `oasis_device.cpp`, which is more complete, and not leave too much words in test comment.

Also, the `write()` function in OASIS will return a list of bytes that represent the bytes in OASIS file. I write a simple script to check it works. The test script is not include in this PR. The script looks like:

```py
import modmesh

device = modmesh.OasisDevice()

rec_poly = modmesh.OasisPolyRecord([
    [70, 720], [410, 720], [410, 920], [70, 920],
    [70, 880], [370, 880], [370, 760], [70, 760]])
rec_rect = modmesh.OasisRectRecord([70, 800], 180, 40)

device.add_poly_record(rec_poly)
device.add_rect_record(rec_rect)

rec_poly = modmesh.OasisPolyRecord([
    [70, 980], [410, 980], [410, 1180], [70, 1180],
    [70, 1140], [370, 1140], [370, 1020], [70, 1020]])
rec_rect = modmesh.OasisRectRecord([70, 1060], 180, 40)

device.add_poly_record(rec_poly)
device.add_rect_record(rec_rect)

with open("test.oas", "wb") as f:
    f.write(bytearray(device.write()))
```

It will generate `test.oas` file. The file in KLayout will looks like:

image

Please review it. Thanks.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+void append_record_bytes(std::vector<uint8_t> bytes, T record);
+
+/**
+ * \class PolyRecord
+ * \brief Convert Rect information to OASIS polygon record bytes.
+ */
+class PolyRecord
+{
+private:
+    std::vector<std::pair<int, int>> vertexes;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: All non-POD class member data need to prefix `m_`.

Put member data after member functions like other modmesh classes do.  Follow convention unless you have a wonderful reason not to.

Use "vertices" as the plural form of "vertex".  Do not use vertexes.

I dislike `std::vector` member data.  But it can be here for a while.  They should be changed to a pad soon.

File: cpp/modmesh/oasis/oasis_device.hpp (hunk 13 above)
Comment (yungyuc) [reply cpp/modmesh/oasis/oasis_device.hpp]: @c1ydehhx You forgot to fix it.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    OasisDevice(OasisDevice const &) = default;
+    OasisDevice(OasisDevice &&) = default;
+    OasisDevice & operator=(OasisDevice const &) = default;
+    OasisDevice & operator=(OasisDevice &&) = default;
+    ~OasisDevice() = default;
+
+    void add_poly_record(const PolyRecord & record);
+    void add_rect_record(const RectRecord & record);
+
+    std::vector<uint8_t> write();
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: If returning a byte container, rename `write` to `to_bytes`.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+static void append_magic_bytes(std::vector<uint8_t> & segment);
+
+template <typename T>
+void append_record_bytes(std::vector<uint8_t> bytes, T record);
+
+/**
+ * \class PolyRecord
+ * \brief Convert Rect information to OASIS polygon record bytes.
+ */
+class PolyRecord
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Rename `PolyRecord` to `OasisRecordPoly`.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    ~PolyRecord() = default;
+
+    std::vector<uint8_t> to_bytes() const;
+};
+
+/**
+ * \class RectRecord
+ * \brief Convert Rect information to OASIS rectangle record bytes.
+ */
+class RectRecord
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Rename `RectRecord` to `OasisRecordRect`.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+#include <utility>
+#include <vector>
+
+namespace modmesh
+{
+
+namespace oasis
+{
+
+static void append_signed_integer(std::vector<uint8_t> & segment, int value);
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: After removing the namespace `modmesh::oasis`, move these free functions to `OasisDevice` as static functions..

File: tests/test_oasis_device.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+
+import unittest
+import modmesh
+
+
+class OasisRectRecordTC(unittest.TestCase):
```
Comment (yungyuc) [inline tests/test_oasis_device.py]: Good one.  Need to be renamed with the C++ class.

File: tests/test_oasis_device.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+    def test_to_byte(self):
+        rec = modmesh.OasisRectRecord([70, 800], 180, 40)
+
+        expected_record_bytes = '\x14\x7B\x00\x00\xB4\x01\x28\x8C\x01\xC0\x0C'
+        record_bytes = rec.to_bytes()
+
+        self.assertEqual(record_bytes, list(map(ord, expected_record_bytes)))
+
+
+class OasisPolyRecordTC(unittest.TestCase):
```
Comment (yungyuc) [inline tests/test_oasis_device.py]: Good one.  Need to be renamed with the C++ class.

File: tests/test_oasis_device.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        expected_record_bytes = '\x15\x3B\x00\x00\x00\x07' \
+                                '\xA8\x05\x90\x03\xA9\x05' \
+                                '\x51\xD8\x04\xF1\x01\xD9' \
+                                '\x04\x8C\x01\xA0\x0B'
+        record_bytes = rec.to_bytes()
+
+        self.assertEqual(record_bytes, list(map(ord, expected_record_bytes)))
+
+
+# For the format of OASIS, please refer OasisDevice comment in oasis_device.cpp
```
Comment (yungyuc) [inline tests/test_oasis_device.py]: Make it clear with the namespace `modmesh::OasisDevice`.

**Review by yungyuc — CHANGES_REQUESTED** [review]

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/base.hpp>
+#include <pybind11/stl.h>
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Do not import any Python-related (pybind11 included) headers outside a `pymod/` directory.

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] Do not include any Python-related (pybind11 included) headers outside a `pymod/` directory.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. Thanks for the review.

I already fix all issues in comment. I adjust the function name, class name, member name, and comment that mention the class name,. Moreover, I remove pybind header that include in wrong place and remove the `oasis` namespace.

For the comment that move free function to `OasisDevice` class, since all classes depends `append_signed_integer`, and `append_unsigned_integer` function, I can't move them into `OasisDevice` class. However, `append_magic_bytes`, `append_start_record_bytes`, `append_cell_and_cell_name_record_byte` and `append_end_record_byte` only used in `OasisDevice`, I move them to `OasisDevice` class.

Please review it. Thanks.

**Comment by yungyuc** [comment]
> For the comment that move free function to `OasisDevice` class, since all classes depends `append_signed_integer`, and `append_unsigned_integer` function, I can't move them into `OasisDevice` class.

You can.  Move `OasisDevice` to be before the declaration of `OasisRecord*`, and forward declare `OasisRecord*` before `OasisDevice`.  See:

```cpp

// Forward declaration (but do not add comment in your PR because it's obvious.
OasisRecordPoly;
OasisRecordRect;

class OasisDevice
{

public:

  // all static functions for OasisRecord*.

private:

  std::vector<OasisRecordPoly> //...;
  std::vector<OasisRecordRect> //...;

};

class OasisRecordPoly
{
  // content.s
};

class OasisRecordRect
{
  // content.s
};
```

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+    std::vector<uint8_t> to_bytes();
+
+private:
+    static void append_magic_bytes(std::vector<uint8_t> & segment);
+    static void append_start_record_bytes(std::vector<uint8_t> & segment);
+    static void append_cell_and_cell_name_record_byte(std::vector<uint8_t> & segment);
+    static void append_end_record_byte(std::vector<uint8_t> & segment);
+
+    template <typename T>
+    static void append_record_bytes(
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: It's too short to use two lines.  Use a single line for the signature.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+    OasisRecordRect(OasisRecordRect const &) = default;
+    OasisRecordRect(OasisRecordRect &&) = default;
+    OasisRecordRect & operator=(OasisRecordRect const &) = default;
+    OasisRecordRect & operator=(OasisRecordRect &&) = default;
+    ~OasisRecordRect() = default;
+
+    std::vector<uint8_t> to_bytes() const;
+
+private:
+    std::pair<int, int> m_bottom_left;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Split the field `m_bottom_left` to be `m_lower` and `m_left`.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+    int w;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: `m_` prefix.  Please review carefully.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+    int h;
+};
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: Class end marker.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+ * \class PolyRecord
+ * \brief Convert Rect information to OASIS polygon record bytes.
+ */
+class OasisRecordPoly
+{
+
+public:
+    explicit OasisRecordPoly(std::vector<std::pair<int, int>> vertices)
+        : m_vertices(std::move(vertices)) {};
+    OasisRecordPoly(OasisRecordPoly const &) = default;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp:95]: You forgot to delete default constructor.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+class OasisRecordRect
+{
+public:
+    OasisRecordRect(std::pair<int, int> bottom_left, int w, int h)
+        : m_bottom_left(bottom_left)
+        , w(w)
+        , h(h)
+    {
+    }
+    OasisRecordRect(OasisRecordRect const &) = default;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp:122]: Ditto.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Points to address:

- [x] Split the field `OasisRecordRect::m_bottom_left` to be `m_lower` and `m_left`.
- [x] Add end marker to all classes.  You forgot some.
- [x] Missing rule of five.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. Thanks for the review.

I fix all issue mentioned in comment. I delete copy assignment on `OasisDevice` and record classes, add all end class marker and adjust signature to one-line, and split `m_bottom_left` to `m_lower` and `m_left`. Also, I check the member of non-POD class should have `m_` prefix again. Sorry for the misunderstand of your instruction.

Please review it. Thanks.

**Review by yungyuc — CHANGES_REQUESTED** [review]
You have previous points unaddressed:

- [x] You forgot to move free functions.
- [x] Missing rule of five.

File: cpp/modmesh/oasis/oasis_device.hpp (hunk 18 above)
Comment (yungyuc) [reply cpp/modmesh/oasis/oasis_device.hpp]: You forgot to move the free functions.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. Thanks for the review.

I just move the free function into `OasisDevice` class, and call these function in `OasisRecord*` class with static way. It looks like:

```cpp
std::vector<uint8_t> OasisRecordRect::to_bytes(){
    // Implementation ...
    OasisDevice::append_unsigned_integer(segment, m_w);
    OasisDevice::append_unsigned_integer(segment, m_h);
    OasisDevice::append_signed_integer(segment, m_left);
    OasisDevice::append_signed_integer(segment, m_lower);
}
```

Also, I delete default constructor in `OasisRecord*`. I previous concern that delete default constructor will happen compilation error since the vector with `OasisRecord*` can not be construct. 

```cpp
    std::vector<OasisRecordPoly> m_polygon_records;
    std::vector<OasisRecordRect> m_rect_records;
```

After think a while, since I don't pre-allocate the space to let the vector using default constructor to create element, it still work properly.

I already fixed these issue. Thanks.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+    OasisDevice(OasisDevice const &) = default;
+    OasisDevice(OasisDevice &&) = default;
+    OasisDevice & operator=(OasisDevice const &) = default;
+    OasisDevice & operator=(OasisDevice &&) = default;
+    ~OasisDevice() = default;
+
+    void add_poly_record(const OasisRecordPoly & record);
+    void add_rect_record(const OasisRecordRect & record);
+
+    static void append_signed_integer(std::vector<uint8_t> & segment, int value);
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: I usually move the static functions to the beginning of the class because they do not depend on other code in it.

It also makes sense to move the static member functions to be before other member functions, because the latter (or the code in the latter) may call the former.

My order is usually: (1) aliases, constants, and enum, (2) static member functions, (3) member functions, (4) member data.  It's also the convention used in modmesh.

Please follow the convention.

**Review by yungyuc — CHANGES_REQUESTED** [review]
> Also, I delete default constructor in `OasisRecord*`. I previous concern that delete default constructor will happen compilation error since the vector with `OasisRecord*` can not be construct.
> 
> ```c++
>     std::vector<OasisRecordPoly> m_polygon_records;
>     std::vector<OasisRecordRect> m_rect_records;
> ```
> 
> After think a while, since I don't pre-allocate the space to let the vector using default constructor to create element, it still work properly.
> 
> I already fixed these issue. Thanks.

If you want to keep the default constructor, make it explicit like:

```cpp
OasisRecordPoly() = default;
```

Otherwise it's deleted implicitly: https://stackoverflow.com/a/38995677.

The point is to make your intention explicit.  You did not seem to know that not supplying the default constructor actually removed it.  It will confuse you and maintainers in the future.

- [x] Review if you want to have the default constructor of `OasisRecord*` classes or not.  If yes, change the code.  If not, leave a comment and stay with the deletion.

Other points to address:

- [x] Move the static functions to the beginning of the class because they do not depend on other code in it.

File: cpp/modmesh/oasis/oasis_device.hpp [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+    OasisRecordRect(OasisRecordRect const &) = default;
+    OasisRecordRect(OasisRecordRect &&) = default;
+    OasisRecordRect & operator=(OasisRecordRect const &) = default;
+    OasisRecordRect & operator=(OasisRecordRect &&) = default;
+    ~OasisRecordRect() = default;
+
+    std::vector<uint8_t> to_bytes() const;
+
+private:
+    int m_left;
```
Comment (yungyuc) [inline cpp/modmesh/oasis/oasis_device.hpp]: You may use 4 lines for the 4 fields or 1 line for them.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. Thanks for the review.

I already fix the function order of class. If my realize correct, the member function may include constructor, so I put static function before constructor. I would like to check is it fit the convention. Thanks. 

For the private static function, I'm not sure the convention that also need to move to the beginning of the class. If so, it will look like this:

```cpp
class OasisDevice
{
public:
    static void append_signed_integer(std::vector<uint8_t> & segment, int value);
    static void append_unsigned_integer(std::vector<uint8_t> & segment, int value);

private:
    static void append_magic_bytes(std::vector<uint8_t> & segment);
    static void append_start_record_bytes(std::vector<uint8_t> & segment);
    static void append_cell_and_cell_name_record_byte(std::vector<uint8_t> & segment);
    static void append_end_record_byte(std::vector<uint8_t> & segment);

public:
    // Public member function.

private:
    // Private member function.
    // Private member.
} /* end class OasisDevice */
```

Please let me know that I just keep the current version or adjust to the version above.

Moreover, I condense rect member to 1 line. 

> Review if you want to have the default constructor of OasisRecord* classes or not. If yes, change the code. If not, leave a comment and stay with the deletion.

After think of the while, I think I'll keep default constructor deleted. Since the empty record of poly or rect is nonsense afaik, I'll keep the default constructor deleted. Thanks for the instruction about the behavior of constructor.

*These changes already co-work with Codex to check the point you address is met.*

**Review by yungyuc — APPROVED** [review]
LGTM

> Please let me know that I just keep the current version or adjust to the version above.

For private static functions, either works.

> After think of the while, I think I'll keep default constructor deleted. Since the empty record of poly or rect is nonsense afaik, I'll keep the default constructor deleted. Thanks for the instruction about the behavior of constructor.

Not necessarily.  We usually reserve in containers.  But maybe reserving can use placement new.  It would be interesting to check what the standard says about it.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please update commit logs and the PR leading comment with a concise summary of what you do.

You do not have to squash although you may.  Request for the next review once it's done.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. I alreday update PR description and squash the commit and add properly message.

Please merge it after CI passed. Thanks.

**Comment by yungyuc** [comment]
> Hi @yungyuc. I alreday update PR description and squash the commit and add properly message.
> 
> Please squash it after CI passed. Thanks.

You already squashed.  But the commit log does not have a clear summary.  This is what I see:

> Add OASIS device that store rect/poly and convert to OASIS format bytes

Please try again.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. I already rewrite the commit message that try to let it more clear. Please review it. Thanks.

**Comment by yungyuc** [comment]
> Hi @yungyuc. I already rewrite the commit message that try to let it more clear. Please review it. Thanks.

You wrote a short commit log but it's not clear.  Try this:

> Add a helper class to return a byte sequence in the OASIS file format.
> 
> * A new class, `modmesh::OasisDevice`, is added.  The member function `to_bytes()` returns the byte sequence.  It currently supports two OASIS record types: polygon and rectangle.
>   * The returned byte sequence can be written to a file.
> * 2 new classes `modmesh::OasisRecordPoly` and `modmesh::OasisRecordRect` correspond to the supported OASIS record types.
> * The 3 new classes are wrapped to Python for testing.

Please supplement missing information and rewrite to make it more clear and concise.

File: cpp/modmesh/oasis/oasis_device.cpp [hunk 31]
```diff
@@ -0,0 +1,315 @@
...
+{
+    // Magic byte should be %SEMI-OASIS<CR><NL>, which <CR><NL> is 0x0D 0x0A.
+    // Please refer OASIS Draft section 6.4.
+
+    std::string magic_byte_plain_text = "%SEMI-OASIS\x0D\x0A";
+
+    for (char c : magic_byte_plain_text)
+    {
+        segment.push_back(c);
+    }
```
Comment (tigercosmos) [inline cpp/modmesh/oasis/oasis_device.cpp]: why not copy the whole buffer directly?

Comment (tigercosmos) [reply cpp/modmesh/oasis/oasis_device.cpp]: same for all other similar operations.

Comment (c1ydehhx) [reply cpp/modmesh/oasis/oasis_device.cpp]: I think it will be better if I copy whole buffer directly. Thanks for the suggestion.

**Comment by yungyuc** [comment]
@c1ydehhx Are you ready for updating the commit log?

**Comment by yungyuc** [comment]
@Han-Xuan ***@***.***>,

Do you have a plan for finishing the commit log?

Regards,
Yung-Yu

From: Yung-Yu Chen ***@***.***>
Date: Wednesday, February 18, 2026 at 14:31
To: solvcon/modmesh ***@***.***>
Cc: Yung-Yu Chen ***@***.***>, Your activity ***@***.***>
Subject: Re: [solvcon/modmesh] Implement OASIS file writer (PR #674)

[https://avatars.githubusercontent.com/u/399122?s=20&v=4]yungyuc left a comment (solvcon/modmesh#674)<https://github.com/solvcon/modmesh/pull/674#issuecomment-3918992777>

@c1ydehhx<https://github.com/c1ydehhx> Are you ready for updating the commit log?

File: cpp/modmesh/oasis/oasis_device.cpp (hunk 31 above)
Comment (c1ydehhx) [reply cpp/modmesh/oasis/oasis_device.cpp]: Hi @tigercosmos. I changes few byte append operation with for-loop with `std::vector` insert operation instead, which makes the code clear. For the other byte append operation with `std::vector::push_back` function, I would like to append bytes sepreately and with comments to clarify how the OASIS format is organized.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. Thanks for the review. I've been updated the commit message with short description summarizing what this PR does.

**Comment by yungyuc** [comment]
> Hi @yungyuc. Thanks for the review. I've been updated the commit message with short description summarizing what this PR does.

Thank you, @c1ydehhx .  It's still unclear and please improve.

**Comment by c1ydehhx** [comment]
Hi @yungyuc. Thanks for the review. I've updated the lead comment and commit message. Please review it. Thanks.

**Review by yungyuc — APPROVED** [review]
LGTM
